fix: handle wasteChuteLocation in frontend

sfoster1 opened the pull request:
This is a new kind of legacy location added in https://github.com/Opentrons/opentrons/pull/19684 (
b801f7f5fda47a36a7b9fbc869fa002a9701a8bc ) to present better information when a protocol tries to use a labware that has been thrown away. Add a type binding for it in the LabwareLocation type that PE load commands
can take and emit, and some ts type predicates for figuring out when it's present.

This is a whole lot of code across the entire FE. it's validated by ts but I need to add tests throughout and do some manual tests.

The PR is organized so each commit is a different subsystem, which might be a nicer way to read it.

## Todo
- [x] update unit tests
- [x] test manually with associated protocol (anything that throws a labware away)

Closes RQA-4772

sfoster1 committed:
feat(shared-data): add wasteChuteLocation

This is a new kind of legacy location added in
https://github.com/Opentrons/opentrons/pull/19684 (
b801f7f5fda47a36a7b9fbc869fa002a9701a8bc ) to present better information
when a protocol tries to use a labware that has been thrown away. Add a
type binding for it in the LabwareLocation type that PE load commands
can take and emit, and some ts type predicates for figuring out when
it's present.

sfoster1 committed:
fix(step-generation): handle wasteChuteLocation

Use the new type predicates to make sure we count wasteChuteLocation as
off deck along with offDeck and systemLocation.

sfoster1 committed:
fix(protocol-designer): handle wasteChuteLocation

Use the new type predicates to make sure we count wasteChuteLocation as
off-deck along with offDeck and systemLocation.

sfoster1 committed:
fix(components): handle wasteChuteLocation

Use the new type predicates to make sure we handle wasteChuteLocation.
Also do a special ugly thing for the labware display locations.

sfoster1 committed:
fix(app): handle wasteChuteLocation

Use the new type predicates to make sure we count wasteChuteLocation as
off deck like offDeck and systemLocation.

sfoster1 committed:
lint

sfoster1 committed:
add some tests for stuff that has them

smb2268 commented on app/src/organisms/ErrorRecoveryFlows/hooks/useDeckMapUtils.ts:
```diff
@@ -445,11 +446,7 @@ export function getSlotNameAndLwLocFrom(
-    location === 'offDeck' ||
-    location === 'systemLocation'
-  ) {
+  if (location == null || !locationIsOnDeck(location)) {
```
nice

smb2268 commented on app/src/organisms/LegacyLabwarePositionCheck/utils/labware.ts:
```diff
@@ -174,7 +170,7 @@ export const getLabwareIdsInOrder = (
         ...labwareLocations.reduce<LabwareToOrder[]>((innerAcc, loc) => {
           let slot = ''
-          if (loc === 'offDeck' || loc === 'systemLocation') {
+          if (!locationIsOnDeck(loc)) {
```
Annoying that we have to update these legacy LPC flows bc they only exist for OT2 which does not have a waste chute, but alas

smb2268 commented on components/src/organisms/CommandText/useCommandTextString/utils/getLabwareDisplayLocation.tsx:
```diff
@@ -65,12 +66,12 @@ export function getLabwareDisplayLocation(
-  if (slotName === 'offDeck' || slotName === 'systemLocation') {
+  if (
+    changeAnyUseOfMeToPreserveStructure_thisIsAnOffDeckLocationInASlotName(
+      slotName
```
lol

smb2268 reviewed (approved):
This looks good to me! I don't know if we're actually trying to render these labware or this copy anywhere but the test coverage looks good enough that I think we're okay to merge

sfoster1 committed:
fix: !on-deck != off-deck

For instance, something on another labware or on a module is
not (directly) on-deck.

sfoster1 committed:
chore: exclude tsc output from vitest

Because then we'll run tests twice.

sfoster1 committed:
lint
